File: aten/src/THNN/generic/MarginCriterion.c
#ifndef TH_GENERIC_FILE
#define TH_GENERIC_FILE "generic/MarginCriterion.c"
#else

void THNN_(MarginCriterion_updateOutput)(
          THNNState *state,
          THTensor *input,
          THTensor *target,
          THTensor *output,
          bool sizeAverage,
          accreal margin_)
{
  scalar_t margin = TH_CONVERT_ACCREAL_TO_REAL(margin_);
  THNN_CHECK_NELEMENT(input, target);
  THNN_CHECK_DIM_SIZE(output, 1, 0, 1);
  scalar_t sum = 0;

  TH_TENSOR_APPLY2(scalar_t, input, scalar_t, target,
    scalar_t z = (margin - *input_data * *target_data);
    sum += z>0 ? z : 0;
  );

  if (sizeAverage)
    sum /= THTensor_(nElement)(input);

  THTensor_(set1d)(output, 0, sum);
}

void THNN_(MarginCriterion_updateGradInput)(
          THNNState *state,
          THTensor *input,
          THTensor *target,
          THTensor *gradInput,
          bool sizeAverage,
          accreal margin_)
{
  scalar_t margin = TH_CONVERT_ACCREAL_TO_REAL(margin_);
  THNN_CHECK_NELEMENT(input, target);
  scalar_t norm = (sizeAverage ? 1./((scalar_t)THTensor_(nElement)(input)) : 1.);

  THTensor_(resizeAs)(gradInput, input);
  TH_TENSOR_APPLY3(scalar_t, gradInput, scalar_t, input, scalar_t, target,
    *gradInput_data = (*input_data * *target_data) < margin ? -norm * *target_data : 0;
  );
}

#endif

File: torch/legacy/nn/MarginCriterion.py
import torch
from .Criterion import Criterion


class MarginCriterion(Criterion):

    def __init__(self, margin=1, sizeAverage=True):
        super(MarginCriterion, self).__init__()
        self.sizeAverage = True
        self.margin = margin
        self.output_tensor = None

    def updateOutput(self, input, target):
        if self.output_tensor is None:
            self.output_tensor = input.new(1)
        self._backend.MarginCriterion_updateOutput(
            self._backend.library_state,
            input,
            target,
            self.output_tensor,
            self.sizeAverage,
            self.margin
        )
        self.output = self.output_tensor[0].item()
        return self.output

    def updateGradInput(self, input, target):
        self._backend.MarginCriterion_updateGradInput(
            self._backend.library_state,
            input,
            target,
            self.gradInput,
            self.sizeAverage,
            self.margin
        )
        return self.gradInput
